cpp/src/arrow/flight/internal.cc

```diff
@@ -49,27 +50,153 @@ Status FromGrpcStatus(const grpc::Status& grpc_status) {
     return Status::OK();
   }
 
-  if (grpc_status.error_code() == grpc::StatusCode::UNIMPLEMENTED) {
-    return Status::NotImplemented("gRPC returned unimplemented error, with message: ",
-                                  grpc_status.error_message());
-  } else {
-    return Status::IOError("gRPC failed with error code ", grpc_status.error_code(),
-                           " and message: ", grpc_status.error_message());
+  StatusCode code = StatusCode::UnknownError;
+  std::string message = "";
+  std::shared_ptr<StatusDetail> detail = nullptr;
+
+  switch (grpc_status.error_code()) {
+    case grpc::StatusCode::OK:
+      return Status::OK();
+    case grpc::StatusCode::CANCELLED:
+      code = StatusCode::IOError;
+      message = util::StringBuilder("gRPC cancelled call, with message: ",
+                                    grpc_status.error_message());
+      detail = std::make_shared<FlightStatusDetail>(FlightStatusCode::Cancelled);
+      break;
+    case grpc::StatusCode::INVALID_ARGUMENT:
+      code = StatusCode::Invalid;
+      message =
+          util::StringBuilder("gRPC returned invalid argument error, with message: ",
+                              grpc_status.error_message());
+      break;
+    case grpc::StatusCode::DEADLINE_EXCEEDED:
+      code = StatusCode::IOError;
+      message =
+          util::StringBuilder("gRPC returned deadline exceeded error, with message: ",
+                              grpc_status.error_message());
+      detail = std::make_shared<FlightStatusDetail>(FlightStatusCode::TimedOut);
+      break;
+    case grpc::StatusCode::NOT_FOUND:
+      code = StatusCode::IOError;
+      message = util::StringBuilder("gRPC returned not found error, with message: ",
+                                    grpc_status.error_message());
+      break;
+    case grpc::StatusCode::ALREADY_EXISTS:
+      code = StatusCode::Invalid;
```

Comment:
I believe I added an already exists error code.

Reply:
Thanks for the pointer, I've changed that.

Do you think NOT_FOUND should also be promoted to the general status codes, as the counterpart?

Reply:
I think KeyError might be the equivalent.

Reply:
Ah I see, thanks - updated.